### ISSUE TITLE
It is a error in VB when I declare a Constant of type Some Class = Nothing

### ISSUE THREAD

**Issue by paul1956** [dotnet/vblang#287]
Version Used:
15.6.4
Steps to Reproduce:
```
Imports System
Namespace Test
    Friend MustInherit Partial Class ConstantValue
        Implements IEquatable(Of ConstantValue)
        Public Const NotAvailable As ConstantValue = Nothing
    End Class
End Namespace
```
**Expected Behavior**:
No errors when setting NotAvailable to Nothing. This works in C# and the IL allows it. or if I declare in a Module NotAvailable ReadOnly. I would also expect this to work for Structure, and other places it works in C# and is allowed in the IL.
**Actual Behavior**:
BC30424 Constants must be of an intrinsic or enumerated type, not a class, structure, type parameter, or array type.

**Comment by rskar-git** [comment]
I can see why there may be some interest in allowing for `Const MyClassNull As MyClass = Nothing` (after all, C# allows it). What I don't see is the point of it - except possibly as syntactic sugar for duck typing? Is that what this is for, nicer duck typing?

**Comment by paul1956** [comment]
I had never heard of duck typing but I guess that is how it is being used, I want it because it is frequently used by C# even in Roslyn so it is a coding style I want to be able to emulate in VB.

**Comment by rskar-git** [comment]
My bad, duck typing is the wrong term here. The term I should have said is inferred typing. 

Anyway, somewhere in someone's coding, inside some method, presumably in a subclass, instead of

    Dim foo As ConstantValue = Nothing

you can now do (assuming `Option Infer On` is in effect)

    Dim foo = NotAvailable

As an aside, duck typing is about letting type checking happen at runtime. VB has allowed that sort of thing like forever (just `Dim` as `Object`); C# got it with `dynamic`.

**Comment by paul1956** [comment]
Unfortunately our standard is Infer Off, Strict On, we find it catches more bugs at design time instead of runtime. Maybe if we had local options this might be a good place for them. The original question is why is this not allowed. "Nothing" should always be allowed as a const.

**Comment by paul1956** [comment]
@rskar-git I hadn't thought about the fact that the pattern below is all over my code. It would be nice to clean it up.

```
FooBar(DirectCast(Nothing, ConstantValue))
```

**Comment by reduckted** [comment]
How interesting. Here's a minimal version of the original code, and the equivalent C# code which compiles:

```vb
Class ConstantValue
    Public Const NotAvailable As ConstantValue = Nothing
                                '~~~~~~~~~~~~~ BC30424
End Class
```

```c#
class ConstantValue {
    public const ConstantValue NotAvailable = null;
}
```

Interestingly, the constant can be declared as `Object`, but is obviously useless, since you don't want the constant to be of type `Object`.

This behaviour _seems_ to be in line with the spec, though the spec is possibly a little bit contradictory. On the one hand it does _not_ state that a constant can be a class type, so the compile error seems to be correct, but the spec also states that a constant can be set to the value of `Nothing`.

> **[VB Constant Expressions](https://github.com/dotnet/vblang/blob/master/spec/expressions.md#constant-expressions)**
> The type of a constant expression can be Byte, SByte, UShort, Short, UInteger, Integer, ULong, Long, Char, Single, Double, Decimal, Date, Boolean, String, Object, or any enumeration type. The following constructs are permitted in constant expressions:
> * Literals (including Nothing).

This is _pretty much_ the same as the C# spec:

> **[C# Constant Expressions](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#constant-expressions)**
> A constant expression must be the null literal or a value with one of the following types: sbyte, byte, short, ushort, int, uint, long, ulong, char, float, double, decimal, bool, object, string, or any enumeration type. Only the following constructs are permitted in constant expressions:
> * Literals (including the null literal).

Given this, I can't see a reason _not_ to allow a constant to be set to `Nothing`.

**Comment by paul1956** [comment]
@reduckted  Is this a simple fix and what would it take to be accepted?

**Comment by reduckted** [comment]
@paul1956 I have absolutely no idea :grin:

**Comment by KathleenDollard** [comment]
@paul1956 

In your code, what about using:

```vb
Public Shared ReadOnly NotAvailable As ConstantValue = Nothing
```

It would seem to serve the same purpose.

**Comment by paul1956** [comment]
Yes ReadOnly has been my workaround in this specific case, it is a little more complicated because this is generated code and I need to deal with Class vs. Module and other corner cases. The original question is why is it not allowed (the spec seems to imply it should work, (as pointed out by @rskar-git ) and why does it work with Infer On or ReadOnly. This is low priority compared to other things.

**Comment by KathleenDollard** [comment]
I wanted to understand the priority. 

Glad to hear this is low priority and an annoyance.
